File: Golang_src/medium/countingBits_338.go
package medium

// Use Bit Manipulation:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 4 ms, faster than 92.41%
// Memory Usage: 4.4 MB, less than 33.79%
func countBits(num int) []int {
	result := make([]int, num+1)

	for i := 1; i <= num; i++ {
		result[i] = result[i>>1] + i&1
	}

	return result
}

File: Cpp_src/medium/counting_bits_338.cc
#include <iostream>
#include <vector>
using namespace std;

// Use bit manipulation
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 52 ms, faster than 94.35%
// Memory Usage: 9.5 MB, less than 82.93%
class Solution
{
public:
    vector<int> countBits(int num)
    {
        vector<int> result(num + 1, 0);
        for (int i = 1; i <= num; i++)
        {
            result[i] = result[i & (i - 1)] + 1;
        }

        return result;
    }
};